### PR TITLE
Create Charm Architecture Reference

### PR DESCRIPTION
**Pull request by addyess**
* Includes documentation of the overview of the charm architecture
* Includes a diagram of users and interactions within the charm model

### CODE REVIEW

File: docs/src/reference/architecture.md
Comment (bschimke95) [inline docs/src/reference/architecture.md:1]: Does it make sense to split the snap and charm architecture?
@evilnick wdyt?

**Review by bschimke95 — APPROVED** [review]
Great write-up, @addyess!

**Comment by bschimke95** [comment]
Any blockers for this PR? It is already open for two weeks.

**Comment by addyess** [comment]
> Any blockers for this PR? It is already open for two weeks.

I think we were looking for a review from @evilnick

File: docs/src/reference/architecture.md [hunk 1]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
   programmatically drive the Kubernetes cluster. 
 
+## Canonical K8s charms
+
+Canonical `k8s` Charms encompass two primary components: the [`k8s` charm][K8s charm] and the [`k8s-worker` charm][K8s-worker charm].
+
+```{kroki} ../assets/charms-architecture.puml
+```
+
+Charms are instantiated on a machine as a juju unit, and a collection of units
```
Comment (evilnick) [inline docs/src/reference/architecture.md:111]:
```suggestion
Charms are instantiated on a machine as a Juju unit, and a collection of units 
```

File: docs/src/reference/architecture.md [hunk 2]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
+
+Canonical `k8s` Charms encompass two primary components: the [`k8s` charm][K8s charm] and the [`k8s-worker` charm][K8s-worker charm].
+
+```{kroki} ../assets/charms-architecture.puml
+```
+
+Charms are instantiated on a machine as a juju unit, and a collection of units 
+constitutes an application. Both `k8s` and `k8s-worker` units are responsible
+for installing and managing its machine's `k8s` snap, however the charm type determines
+the node's role in the kubernetes cluster. The `k8s` charm manages `control-plane` nodes,
```
Comment (evilnick) [inline docs/src/reference/architecture.md:114]:
```suggestion
the node's role in the Kubernetes cluster. The `k8s` charm manages `control-plane` nodes,
```

File: docs/src/reference/architecture.md [hunk 3]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
+whereas the `k8s-worker` charm manages kubernetes `worker` nodes. The administrator manages
```
Comment (evilnick) [inline docs/src/reference/architecture.md:115]:
```suggestion
whereas the `k8s-worker` charm manages Kubernetes `worker` nodes. The administrator manages 
```

File: docs/src/reference/architecture.md [hunk 4]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
+for installing and managing its machine's `k8s` snap, however the charm type determines
+the node's role in the kubernetes cluster. The `k8s` charm manages `control-plane` nodes,
+whereas the `k8s-worker` charm manages kubernetes `worker` nodes. The administrator manages 
+the cluster via the `juju` client, directing the `juju` controller to reach the model's 
+eventually consistent state. See the [juju docs][Juju docs] to understand more juju's concepts.
+
+The administrator may choose any supported cloud-types (Openstack, MAAS, AWS, GCP, Azure...) on
+which to manage the machines making up the Kubernetes cluster. Juju selects a single leader unit
+per application to act as a centralized figure with the model. The `k8s` leader oversees kubernetes 
+bootstrapping and enlisting of new nodes. Follower `k8s` units will join the cluster utilizing
```
Comment (evilnick) [inline docs/src/reference/architecture.md:122]:
```suggestion
bootstrapping and enlistment of new nodes. Follower `k8s` units will join the cluster using
```

File: docs/src/reference/architecture.md [hunk 5]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
+```{kroki} ../assets/charms-architecture.puml
+```
+
+Charms are instantiated on a machine as a juju unit, and a collection of units 
+constitutes an application. Both `k8s` and `k8s-worker` units are responsible
+for installing and managing its machine's `k8s` snap, however the charm type determines
+the node's role in the kubernetes cluster. The `k8s` charm manages `control-plane` nodes,
+whereas the `k8s-worker` charm manages kubernetes `worker` nodes. The administrator manages 
+the cluster via the `juju` client, directing the `juju` controller to reach the model's 
+eventually consistent state. See the [juju docs][Juju docs] to understand more juju's concepts.
```
Comment (evilnick) [inline docs/src/reference/architecture.md:117]:
```suggestion
eventually consistent state. For more detail on Juju's concepts, see the [Juju docs][].
```

File: docs/src/reference/architecture.md [hunk 6]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
+which to manage the machines making up the Kubernetes cluster. Juju selects a single leader unit
+per application to act as a centralized figure with the model. The `k8s` leader oversees kubernetes 
+bootstrapping and enlisting of new nodes. Follower `k8s` units will join the cluster utilizing
+secrets shared through relation data from the leader. The entire lifecycle of the deployment
+is orchestrated by the `k8s` charm, with tokens and cluster-related information being exchanged 
+through Juju relation data.
+
+Furthermore, the `k8s-worker` unit functions exclusively as a worker within the cluster, establishing
+a relation with the `k8s` leader unit and requesting tokens and cluster-related information through
+relation data. The `k8s` leader is responsible for issuing these tokens and revoking them when if
```
Comment (evilnick) [inline docs/src/reference/architecture.md:129]:
```suggestion
relation data. The `k8s` leader is responsible for issuing these tokens and revoking them if
```

File: docs/src/reference/architecture.md [hunk 7]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
...
+a unit administratively departs the cluster.
+
+Moreover, the `k8s` charm supports the integration of other compatible charms, enabling integrations
```
Comment (evilnick) [inline docs/src/reference/architecture.md:132]:
```suggestion
The `k8s` charm also supports the integration of other compatible charms, enabling integrations 
```

**Review by evilnick — APPROVED** [review]
Nice, just a few things to tweak before you claim your :cake:

File: docs/src/reference/architecture.md [hunk 8]
```diff
@@ -101,6 +101,43 @@ This functionality is exposed via the following interfaces:
 - The **API**: The API over HTTP serves the CLI and is also used to
   programmatically drive the Kubernetes cluster. 
 
+## Canonical K8s charms
+
+Canonical `k8s` Charms encompass two primary components: the [`k8s` charm][K8s charm] and the [`k8s-worker` charm][K8s-worker charm].
```
[the remark this answers is not in the document]
Comment (addyess) [reply docs/src/reference/architecture.md:106]: @bschimke95 

Where would  people likely be searching?  On charmhub they are loving titled `k8s` and `k8s-worker`.  This doc links to the charms on charmhub.io.  Do you think a link to the k8s-operator gh repo might be nice here too?

**Pull request merged**